### PR TITLE
#if targetEnvironment(simulator)

### PR DESCRIPTION
**Pull request by graydon**
This is an implementation of https://github.com/apple/swift-evolution/pull/764 which has now been accepted. It is ready to merge.

### CODE REVIEW

**Comment by graydon** [comment]
@swift-ci please test

File: include/swift/AST/DiagnosticsParse.def [hunk 1]
```diff
@@ -1503,6 +1503,9 @@ ERROR(empty_version_string,none,
 WARNING(unknown_platform_condition_argument,none,
       "unknown %0 for build configuration '%1'",
       (StringRef, StringRef))
+WARNING(likely_simulator_platform_condition,none,
+      "plaform condition appears to be testing for simulator environment",
```
Comment (DougGregor) [inline include/swift/AST/DiagnosticsParse.def]: Could you tack on "; use 'targetEnvironment(simulator)' instead" ? That way, one doesn't have to look at the contents of the Fix-It to see what the fix is.

File: lib/Basic/LangOptions.cpp [hunk 2]
```diff
@@ -254,6 +262,13 @@ std::pair<bool, bool> LangOptions::setTarget(llvm::Triple triple) {
   else
     addPlatformConditionValue(PlatformConditionKind::Runtime, "_Native");
 
+  // Set the "targetEnvironment" platform condition if targeting a simulator
+  // environmet. Otherwise _no_ value is present for targetEnvironment; it's
```
Comment (DougGregor) [inline lib/Basic/LangOptions.cpp]: Typo `environmet`

**Review by DougGregor — APPROVED** [review]
This looks fantastic, thank you @graydon

**Comment by graydon** [comment]
@swift-ci please test

File: lib/Parse/ParseIfConfig.cpp [hunk 3]
```diff
@@ -544,8 +547,97 @@ static bool isVersionIfConfigCondition(Expr *Condition) {
   return IsVersionIfConfigCondition().visit(Condition);
 }
 
+/// Get the identifier string from an \c Expr if it's an
+/// \c UnresolvedDeclRefExpr, otherwise the empty string.
+static StringRef getDeclRefStr(Expr *E) {
+  if (auto *UDRE = cast<UnresolvedDeclRefExpr>(E)) {
```
Comment (compnerd) [inline lib/Parse/ParseIfConfig.cpp]: Am I just mistaken, or shouldn't this be a `dyn_cast`?  `cast` implies that this must succeed and will abort if the expression type is not a `UnresolvedDeclRefExpr`.

Comment (graydon) [reply lib/Parse/ParseIfConfig.cpp]: Oh, yes, this is a typo, thanks! Will fix tomorrow.

**Pull request merged**